source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,270 @@ extern "C"
+                         ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                         int wgt_dim, float *obj_wgts, int *ierr)
+    {
+      SparsityPattern *graph = (SparsityPattern *)data;
+      *ierr = ZOLTAN_OK;
+
+      Assert( globalID != NULL , ExcEmptyObject() );
+      Assert( localID != NULL , ExcEmptyObject() );
+
+      //set global degrees of freedom
+      auto n_dofs = graph->n_rows();
+      for (int i=0; i < n_dofs; i++)
+        {
+
+          globalID[i] = i;    //global ids run from 0 to n_dofs-1
+          localID[i] = i;     //same as global ids
+
+        }
+      return;
+    }
+
+    void get_num_edges_list(void *data, int sizeGID, int sizeLID,
+                            int num_obj,
+                            ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                            int *numEdges, int *ierr)
+    {
+
+      SparsityPattern *graph = (SparsityPattern *)data;
+      *ierr = ZOLTAN_OK;
+
+      Assert ( numEdges != NULL , ExcEmptyObject() );
+
+      for (int i=0; i<num_obj; ++i)
+        {
+          numEdges[i] = graph->row_length(globalID[i])-1;
+          // std::cout << "Number of edges of " << globalID[i] << ":" << numEdges[i] << std::endl;
+        }
+
+    }
+
+    void get_edge_list(void *data, int sizeGID, int sizeLID,
+                       int num_obj, ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                       int *num_edges,
+                       ZOLTAN_ID_PTR nborGID, int *nborProc,
+                       int wgt_dim, float *ewgts, int *ierr)
+    {
+
+      SparsityPattern *graph = (SparsityPattern *)data;
+      *ierr = ZOLTAN_OK;
+
+      ZOLTAN_ID_PTR nextNborGID = nborGID;
+      int *nextNborProc = nborProc;
+
+      Assert( nextNborGID != NULL , ExcEmptyObject() );
+      Assert( nextNborProc != NULL , ExcEmptyObject() );
+
+      for (int i=0; i < num_obj; ++i)
+        {
+
+          for ( SparsityPattern::iterator col=graph->begin(i);
+                col < graph->end(i); ++col )
+            {
+
+              if ( i != col->column() )
+                {
+                  *nextNborGID++ = col->column();
+                  //std::cout << "Neighbours " << globalID[i] << " : " << *(nextNborGID-1) << std::endl;
+                  *nextNborProc++ = 0;    //All the vertices on processor 0
+                }
+            }
+
+        }
+
+
+    }
+
+
+    void partition_zoltan (const SparsityPattern     &sparsity_pattern,
+                           const unsigned int         n_partitions,
+                           std::vector<unsigned int> &partition_indices)
+    {
+      float version;
+
+      //Initialize MPI and Zoltan. Make sure MPI is already initialized.
+      Zoltan *zz = new Zoltan(MPI::COMM_SELF);
+
+      //Check if zoltan object is initialized
+      Assert( zz != NULL , ExcEmptyObject() );
+
+      //General parameters
+      zz->Set_Param( "LB_METHOD", "GRAPH" );  //graph based partition method (LB-load balancing)
```

Comment:
Is this the only partitioning method that you plan on using?

Reply:
I chose graph partitioners based on the application and ease of implementation. Even with this one partitioner, there are different packages available ( set by using GRAPH_PACKAGE). I chose the default one.

I have not considered using geometric and simple partitioner (uses no connectivity info.) here. That leaves us with graph partitioners and hyper graph partitioners. In order to use hyper graph partitioners, other query functions have to be written and is in general more expensive than graph option.